src/recommendation/src/main/scala/RecommendationHelper.scala

```diff
@@ -33,6 +40,95 @@ object SparkHelper {
 
 trait RecommendationParams extends ALSParams
 
+trait RecommendationModelParams extends Params with ALSModelParams with HasPredictionCol {
+
+  def getALSModel(uid: String,
+    rank: Int,
+    userFactors: DataFrame,
+    itemFactors: DataFrame): ALSModel = {
+    new ALSModel(uid, rank, userFactors, itemFactors)
+  }
+
+  def recommendForAllItems(k: Int): DataFrame = ???
+
+  def recommendForAllUsers(k: Int): DataFrame = ???
+
+  def recommendForAllItems(rank: Int, userDataFrame: DataFrame, itemDataFrame: DataFrame, k: Int): DataFrame = {
+    getALSModel(uid, rank, userDataFrame, itemDataFrame)
+      .setUserCol($(userCol))
+      .setItemCol($(itemCol))
+      .recommendForAllItems(k)
+  }
+
+  def recommendForAllUsers(
+    alsModel: ALSModel,
+    num: Int): DataFrame =
+    alsModel.recommendForAllUsers(num)
+
+  def recommendForAllItems(
+    alsModel: ALSModel,
+    num: Int): DataFrame =
+    alsModel.recommendForAllItems(num)
+
+  def recommendForAllUsers(userDataFrame: DataFrame, itemDataFrame: DataFrame, k: Int): DataFrame = {
+    import userDataFrame.sparkSession.implicits._
+    val itemRDD = if (true) {
+      itemDataFrame.rdd.flatMap(row =>
+        row.getAs[mutable.WrappedArray[Float]](1).zipWithIndex.map(r => Row(row.getInt(0), r._2, r._1))
+      ).map(item =>
+        MatrixEntry(item.getInt(0).toLong, item.getInt(1).toLong, item.getFloat(2).toDouble))
+    } else {
+      itemDataFrame.sparkSession.read.parquet("./items-parquet")
+        .rdd.flatMap(row => row.getAs[mutable.WrappedArray[Float]](1).zipWithIndex.map(r =>
+        Row(row.getInt(0), r._2, r._1))
+      ).map(item => MatrixEntry(item.getInt(0).toLong, item.getInt(1).toLong, item.getFloat(2).toDouble))
+    }
+    val left = userDataFrame
+
+    val userRDD = left.rdd.flatMap(row => row.getAs[mutable.WrappedArray[Float]](1).zipWithIndex.map(r =>
+      Row(row.getInt(0), r._2, r._1))
+    ).map(item => MatrixEntry(item.getInt(0).toLong, item.getInt(1).toLong, item.getFloat(2).toDouble))
+
+    val userMatrix = new CoordinateMatrix(userRDD).toBlockMatrix().cache()
+    val itemMatrix = new CoordinateMatrix(itemRDD).toBlockMatrix().cache()
+
+    val userToItemMatrix = userMatrix
+      .multiply(itemMatrix)
+      .toIndexedRowMatrix()
+      .rows.map(index => (index.index.toInt, index.vector))
+
+    val sc = userDataFrame.sparkSession
+
+    val zipVector = udf((list: DenseVector) => {
+      val zipped = list.toArray.zipWithIndex
+      val sorted = zipped.map(f => (f._2, f._1)).sortWith(_._2 > _._2)
+      sorted.take(k)
+    })
+
+    val recs = sc.createDataFrame(userToItemMatrix)
+      .toDF("id", "ratings").withColumn("recommendations", zipVector($"ratings")).select("id", "recommendations")
```

Comment:
in general it's best to move these kinds of common strings, especially if they repeat several times, to a constants file - maybe you could add a common constants file for recommenders and put them there